File: 2237-longest-palindrome-by-concatenating-two-letter-words/longest-palindrome-by-concatenating-two-letter-words.java
class Solution {
    public int longestPalindrome(String[] words) {
        Map<String,Integer> hmap = new HashMap<String,Integer>();
        int res = 0;
        for (String word: words) {
            String reversed = new StringBuffer(word).reverse().toString();
            if (hmap.getOrDefault(reversed, 0) > 0) {
                res += 4;
                hmap.put(reversed, hmap.get(reversed) - 1);
            } else {
                hmap.put(word, hmap.getOrDefault(word, 0) + 1);
            }
        }
        for (Map.Entry<String,Integer> entry: hmap.entrySet()) {
            String key = entry.getKey();
            if (key.charAt(0) == key.charAt(1) && entry.getValue() > 0) {
                res += 2;
                break;
            }
        }
        return res;
    }
}

File: 2237-longest-palindrome-by-concatenating-two-letter-words/longest-palindrome-by-concatenating-two-letter-words.py
class Solution:
    def longestPalindrome(self, words: List[str]) -> int:
        res = 0
        hmap = defaultdict(int)
        for word in words:
            if hmap[word[::-1]] > 0:
                res += 4
                hmap[word[::-1]] -= 1
            else:
                hmap[word] += 1
        for k,v in hmap.items():
            if k[0] == k[1] and v:
                res += 2
                break
        return res